### PR TITLE
Replace the TreeMap in the composite aggregation

### PR DESCRIPTION
**Pull request by jimczi**
The `composite` aggregation uses a TreeMap to keep track of the best buckets.
This ensures a log(n) time cost to insert new buckets but also to retrieve buckets
that are already present in the map. In order to speed up the retrieval of buckets
this change replaces the TreeMap with a priority queue and a HashMap. The insertion
cost is still log(n) but the retrieval of buckets through the HashMap is now done in constant
time. This optimization can bring significant improvement since each document needs
to check if its associated buckets are already present in the current best buckets.

### CODE REVIEW

**Comment by jimczi** [comment]
@elasticmachine run gradle build tests 1

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueueTests.java [hunk 1]
```diff
@@ -133,9 +133,9 @@ public void testRandom() throws IOException {
     }
 
     private void testRandomCase(ClassAndName... types) throws IOException {
-        testRandomCase(true, true, types);
+        //testRandomCase(true, true, types);
         testRandomCase(true, false, types);
-        testRandomCase(false, true, types);
+        //testRandomCase(false, true, types);
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueueTests.java]: why is it disabled?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java [hunk 2]
```diff
@@ -116,12 +130,13 @@ public InternalAggregation buildAggregation(long zeroBucket) throws IOException
...
-        int pos = 0;
-        for (int slot : queue.getSortedSlot()) {
+        int pos = queue.size() - 1;
+        while (queue.size() > 0) {
+            int slot = queue.pop();
             CompositeKey key = queue.toCompositeKey(slot);
             InternalAggregations aggs = bucketAggregations(slot);
             int docCount = queue.getDocCount(slot);
-            buckets[pos++] = new InternalComposite.InternalBucket(sourceNames, formats, key, reverseMuls, docCount, aggs);
+            buckets[pos--] = new InternalComposite.InternalBucket(sourceNames, formats, key, reverseMuls, docCount, aggs);
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java]: we don't actually need to track a `pos` variable, we could use `queue.size()` instead?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java [hunk 3]
```diff
@@ -260,7 +275,7 @@ public void collect(int doc, long zeroBucket) throws IOException {
     }
 
     private SingleDimensionValuesSource<?> createValuesSource(BigArrays bigArrays, IndexReader reader, Query query,
-                                                              CompositeValuesSourceConfig config, int sortRank, int size) {
+                                                              CompositeValuesSourceConfig config, int size, int sortRank) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java]: was it a bug?

**Review by jpountz — COMMENTED** [review]
This sounds like a good idea. I'm not fully familiar with the composite agg so I left some questions.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java [hunk 4]
```diff
@@ -73,6 +76,13 @@
                         List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData,
                         int size, CompositeValuesSourceConfig[] sourceConfigs, CompositeKey rawAfterKey) throws IOException {
         super(name, factories, context, parent, pipelineAggregators, metaData);
+        // check that the provided size is not greater than the search.max_buckets setting
+        int limit = context.aggregations().multiBucketConsumer().getLimit();
+        if (size > context.aggregations().multiBucketConsumer().getLimit()) {
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java]: ;), https://github.com/elastic/elasticsearch/pull/36675/commits/b80398874b10bfce242373efe8b7f997ff610694

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java [hunk 5]
```diff
@@ -88,8 +98,12 @@
...
     @Override
     protected void doClose() {
-        Releasables.close(queue);
-        Releasables.close(sources);
+        if (queue != null) {
+            Releasables.close(queue);
+        }
+        if (sources != null) {
+            Releasables.close(sources);
+        }
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java]: good catch, I added a try/finally in https://github.com/elastic/elasticsearch/pull/36675/commits/b80398874b10bfce242373efe8b7f997ff610694

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueueTests.java (hunk 1 above)
Comment (jimczi) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueueTests.java]: it's a leftover, I  enabled these tests in https://github.com/elastic/elasticsearch/pull/36675/commits/b80398874b10bfce242373efe8b7f997ff610694

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java (hunk 2 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java]: ++, I pushed https://github.com/elastic/elasticsearch/pull/36675/commits/b80398874b10bfce242373efe8b7f997ff610694

**Review by jpountz — APPROVED** [review]
LGTM. Nice that it doesn't make things more complicated.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java (hunk 3 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java]: not exactly, this was used to determine if we should use global ordinals or a binary value source for the leading composite source but the code has been refactored and global ordinals are now fast even when they are used as the leading source. I removed this code in https://github.com/elastic/elasticsearch/pull/36675/commits/b80398874b10bfce242373efe8b7f997ff610694 and we now use global ordinals all the time.

**Pull request merged**